Regtest

vicsn opened the pull request:
Adds a regression test network type to allow easier creation of functional and integration tests.

Some notes:

- on_generateblocks RPC call combines functionality from the on_getblocktemplate and on_submitblock RPC calls to allow rapid block creation. Difficulty is set permanently to 1 for regtest.
- In the future, many conditional statements can potentially be replaced by an abstracted class of the networktype, similar to how Bitcoin does this. [See Bitcoin's CChainParams class](https://github.com/bitcoin/bitcoin/blob/master/src/chainparams.cpp)
- I added two small functional tests to give an initial impression of future tests.  [See Bitcoin's functional tests](https://github.com/bitcoin/bitcoin/tree/master/test/functional)

*** edit june 4th 2018 ***

Noteworthy regtest changes include, in order of perceived importance: 
- forced difficulty
- Empty seed peers list
- A low or custom RECENT_OUTPUT_BLOCKS amount to enable transacting at low height

I thought it would be a nice design choice, but for now I have to agree that the following may be better:
- A new forced_difficulty flag
- Functional tests need to use --offline or --add-exclusive-node
- A new recent_output_blocks flag. Making this separate flag seems more clear than making RECENT_OUTPUT_BLOCKS dependent on the forced_difficulty flag.

If agreed, I'll incorporate the flags and undo a lot of changes. I can continue on functional tests in a new PR.

moneromooo-monero commented on src/rpc/core_rpc_server_commands_defs.h:
```diff
@@ -950,9 +950,7 @@ namespace cryptonote
-      bool mainnet;
-      bool testnet;
-      bool stagenet;
+      std::string nettype;
```
This needs bumping RPC version (and it's gratuitously backward incompatible)

vicsn replied:
Good point. But currently the mainnet/stagenet bools of the DaemonInfo object do not get printed. Shall I add a bool stagenet, stick to std::string nettype, or leave it?

moneromooo-monero replied:
A string type seems like a good idea, but it should preserve the booleans for code that expects them.

vicsn replied:
ack

stoffu replied:
Is this field used anywhere?

stoffu replied:
This patch seems irrelevant for the regtest itself at this point, though I'm OK with adding this string field. However, it seems incomplete because there is no corresponding change in core_rpc_server.cpp.

vicsn replied:
Adjusted

moneromooo-monero commented on src/rpc/core_rpc_server_commands_defs.h:
```diff
@@ -1149,6 +1145,33 @@ namespace cryptonote
+    struct request
+    {
+      uint64_t amount_of_blocks;
+      uint64_t reserve_size;
```
That is pointless since it's not used. This is used by pools to embed something miner specific, but here the blocks are generated without adding anything in the reserved space AFAICT.

vicsn replied:
ack

moneromooo-monero commented on tests/README.md:
```diff
@@ -50,6 +50,10 @@ To run the same tests on a release build, replace `debug` with `release`.
 
 [TODO]
+Functional tests are located under the `tests/functional` directory. 
+Requires installing: https://github.com/emesik/monero-python
```
How needed is this ? Needing some random external code to run tests seems dangerous.

vicsn replied:
It is just a small wrapper for rpc calls. I will message the current maintainer to ask whether any written consent is needed to copy it, and otherwise rewrite from scratch.

moneromooo-monero commented on tests/README.md:
```diff
@@ -50,6 +50,10 @@ To run the same tests on a release build, replace `debug` with `release`.
 [TODO]
+Functional tests are located under the `tests/functional` directory. 
+Requires installing: https://github.com/emesik/monero-python
+Make sure you run the monero daemon with rpc port 28081, then start tests by invoking ./blockchain.py or ./speed.py
```
Is this right ? This introduces a whole new network type with a new default port. Presumably these tests use it ?

vicsn replied:
My mistake. If we keep the regtest network type, it should indeed be adjusted to 48081 mentioned in the cryptonote_config.h. If we do not create a new network type, port 28081 is fine as tests can occur in testnet.

moneromooo-monero commented on tests/functional_tests/blockchain.py:
```diff
@@ -0,0 +1,127 @@
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+# THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+# 
+# Parts of this file are originally copyright (c) 2014-2016 The Bitcoin Core developers
```
I assume the licence is OK to reuse like this ?

moneromooo-monero commented on tests/functional_tests/blockchain.py:
```diff
@@ -0,0 +1,127 @@
+Test the following RPCs:
+    - get_info
+
+Tests correspond to code in ...
```
Left unfinished?

vicsn replied:
Added explicit "TODO"

moneromooo-monero commented on tests/functional_tests/blockchain.py:
```diff
@@ -0,0 +1,127 @@
+        assert_equal(sorted(res.keys()), sorted(keys))
+
+        # free_space should be > 0
+        assert_greater_than(res['free_space'], 0)
```
That seems like a bad test, since there's no guarantee technically :)

vicsn replied:
Changed it to ">=" :). Improving the overall test logic is for a later PR

moneromooo-monero commented on tests/functional_tests/test_framework/authproxy.py:
```diff
@@ -0,0 +1,180 @@
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this software; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
```
Not sure we want LGPL code in the repo, even though for tests it should not matter that much. I'll check with others.

hyc replied:
Why is a proxy even needed?

moneromooo-monero commented on tests/functional_tests/test_framework/authproxy.py:
```diff
@@ -0,0 +1,180 @@
+HTTP_TIMEOUT = 30
+USER_AGENT = "AuthServiceProxy/0.1"
+
+log = logging.getLogger("BitcoinRPC")
```
Is this a verbatim copy and should not change ? Will this particular call fail or just log with a "BitcoinRPC" name ?

moneromooo-monero commented on tests/functional_tests/test_framework/util.py:
```diff
@@ -0,0 +1,568 @@
+    """Assert the fee was in range"""
+    target_fee = tx_size * fee_per_kB / 1000
+    if fee < target_fee:
+        raise AssertionError("Fee of %s BTC too low! (Should be %s BTC)" % (str(fee), str(target_fee)))
```
same here about BTCh

moneromooo-monero commented on tests/functional_tests/test_framework/util.py:
```diff
@@ -0,0 +1,568 @@
+
+def assert_fee_amount(fee, tx_size, fee_per_kB):
+    """Assert the fee was in range"""
+    target_fee = tx_size * fee_per_kB / 1000
```
1024, and quantized

moneromooo-monero commented on tests/functional_tests/test_framework/util.py:
```diff
@@ -0,0 +1,568 @@
+    if fee < target_fee:
+        raise AssertionError("Fee of %s BTC too low! (Should be %s BTC)" % (str(fee), str(target_fee)))
+    # allow the wallet's estimation to be at most 2 bytes off
+    if fee > (tx_size + 2) * fee_per_kB / 1000:
```
The wallet estimation may be more than that.
Does bitcoin do kB based fees too, or is this changed for monero ?

vicsn replied:
Removed the file

stoffu commented:
I agree with @moneromooo-monero's point; the big set of changes due to the additional network type just for the purpose of regression testing seems like a bit too much. I've experimented with another minimal approach that activates the regression testing mode with the `--regtest` flag: https://github.com/stoffu/monero/tree/regtest

Please feel free to cherry-pick it or make adjustments as desired.

moneromooo-monero commented:
By the way, there is a "fake" network that's used by the core tests currently. It's not accessible by command line. Maybe that can be used for the regtest (I dont' think you don't want a public history for tests anyway, and even more so for reproducibility).

stoffu commented:
@moneromooo-monero Thanks for the good suggestion. I made use of the `FAKECHAIN` mode thus eliminating the `m_regtest` variable, and now the regression test will use the blockchain data stored under `~/.bitmonero/fake/`. Please see the updated branch.

moneromooo-monero commented:
Does it reset the chain when loading ? The core tests do, which ensure you don't get stale data when you start testing.

stoffu commented:
@moneromooo-monero Thanks for the feedback, I've updated the code accordingly.

vicsn commented:
See the latest 5 commits for my changes and short explanations. If accepted I'll squash the commits into 1) regtest 2) first functional tests. (unless you think otherwise is better)

moneromooo-monero commented on src/cryptonote_core/blockchain.cpp:
```diff
@@ -363,8 +374,8 @@ bool Blockchain::init(BlockchainDB* db, const network_type nettype, bool offline
-    for (size_t n = 0; test_options->hard_forks[n].first; ++n)
-      m_hardfork->add_fork(test_options->hard_forks[n].first, test_options->hard_forks[n].second, 0, n + 1);
+    for (size_t n = 0; n < sizeof(fakechain_hard_forks) / sizeof(fakechain_hard_forks[0]); ++n)
+      m_hardfork->add_fork(fakechain_hard_forks[n].version, fakechain_hard_forks[n].height, fakechain_hard_forks[n].threshold, fakechain_hard_forks[n].time);
```
That is destroying the core tests. The fake chain has custom hard fork points for a reason.

stoffu replied:
Yes, this patch seems unnecessary.

vicsn replied:
I should've ran the core tests! Reverted this

moneromooo-monero commented on src/rpc/core_rpc_server.cpp:
```diff
@@ -1209,6 +1209,43 @@ namespace cryptonote
+
+      res.height = template_res.height;
+      res.status = submit_res.status;
+    }
```
if req.amount_of_blocks is 0, res.status is uninitialized.
The interetsing case where you might want to carry it over is when a RPC returns false anyway, not true.

vicsn replied:
Added proper initialization for height, and added more moments where status is updated

moneromooo-monero commented on src/rpc/core_rpc_server.h:
```diff
@@ -129,6 +129,7 @@ namespace cryptonote
         MAP_JON_RPC_WE("getblocktemplate",       on_getblocktemplate,           COMMAND_RPC_GETBLOCKTEMPLATE)
         MAP_JON_RPC_WE("submit_block",           on_submitblock,                COMMAND_RPC_SUBMITBLOCK)
         MAP_JON_RPC_WE("submitblock",            on_submitblock,                COMMAND_RPC_SUBMITBLOCK)
+        MAP_JON_RPC_WE("generateblocks",         on_generateblocks,             COMMAND_RPC_GENERATEBLOCKS)
```
Maybe make this a restricted command. Doesn't matter a whole lot though since it errors out when not in fake mode.

vicsn replied:
This was a new element for me! Doesn't seem to hurt to add it

moneromooo-monero commented on src/rpc/core_rpc_server_commands_defs.h:
```diff
@@ -1149,6 +1150,31 @@ namespace cryptonote
       END_KV_SERIALIZE_MAP()
     };
   };
```
nettype isn't serialized

vicsn replied:
I should've caught that

moneromooo-monero commented on tests/functional_tests/blockchain.py:
```diff
@@ -0,0 +1,104 @@
+        daemon = Daemon()
+        res = daemon.get_info()
+
+        assert sorted(res.keys()) == sorted(keys)
```
This means every time someone adds something, this has to be updated too, for no other reason that there is a list above. That sounds a bit like busy work maybe. How about just check all keys you are testing (free space and height here) are present instead ?

vicsn replied:
Indeed, adjusted to check space/height individually. Can be abstracted in the future to e.g. always test for existence and a predicate

moneromooo-monero commented on tests/functional_tests/blockchain.py:
```diff
@@ -0,0 +1,104 @@
+        assert res['free_space'] >= 0
+
+        # height should be greater or equal to 0
+        assert res['height'] >= 0
```
Height should never be 0, the genesis block is supposed to be added on init. If not, bug.

vicsn replied:
Adjusted

moneromooo-monero commented on tests/functional_tests/test_framework/wallet.py:
```diff
@@ -0,0 +1,103 @@
+
+    def transfer(self, destinations, ringsize=7, payment_id=''):
+        transfer = {
+            'method': 'transfer',
```
For the record, transfer is kind of obsolete. You want to use transfer_split nowadays.

vicsn replied:
Thanks for the info :)

moneromooo-monero commented on tests/functional_tests/test_framework/wallet.py:
```diff
@@ -0,0 +1,103 @@
+            'jsonrpc': '2.0', 
+            'id': '0'    
+        }
+        if(len(payment_id) > 1):
```
Odd test. Payment IDs are fixed length (8 bytes or 32 bytes, twice those if in hex or course). If you just want to test whether empty, > 0.

vicsn replied:
Adjusted to  > 0

moneromooo-monero replied:
Still shows as > 1 here.

vicsn replied:
Back to 0 again, thanks for spotting. I also noticed I missed updating the "assert res['height'] >= 0". Perhaps because I cherry-picked/rebased in the wrong way, will pay attention to this.

stoffu commented on src/cryptonote_core/blockchain.cpp:
```diff
@@ -419,6 +420,10 @@ bool Blockchain::init(BlockchainDB* db, const network_type nettype, bool offline
+  else if (m_regtest)
+  {
+    m_db->set_batch_transactions(true);
+  }
```
The reg test wouldn’t work if it was omitted IIRC. The core tests also has this call.

moneromooo-monero replied:
That seems odd since there are very few differences between non-core-tests fakechain and mainnet. In any case, I asked hyc whether batch transactions can be enabled by default.

moneromooo-monero replied:
hyc sees no reason why they should not be enabled by default. Let it be so :)

vicsn replied:
Shall I:
- remove the "if (m_regtest)"
- remove "set_batch_transactions(true);" in the other locations: 
blockchain_db.cpp/db_lmdb.cpp/blockchain_import.cpp/hardfork.cpp (and chaingen.h if I don't break anything) ?

moneromooo-monero replied:
I'd remove that whole if (regtest) { batch; } thing, and default batch transactions to true wherever they're defined (db_lmdb.cpp probably). That is, probably set a bool to true in the ctor, wahtever it is the function does.

stoffu replied:
I've pushed my interpretation of @moneromooo-monero's suggestion to my updated branch: https://github.com/stoffu/monero/commits/regtest

stoffu commented:
The regtest really doesn't seem to work unless `--offline` is specified. Have you tried it yourself? As far as I tried some attempts (e.g. changing the network ID), the daemon always finds other mainnet peers and tries to get synced, which makes it always return the "BUSY" response for the `generateblock` RPC.

I think making the regtest work without the offline flag could be worked on later.

vicsn commented:
@stoffu I think it works for me because I always fire up three local nodes which are connected to each other. A single node without any connections will say it is busy (connecting...). It'll be nice to have multiple nodes running when testing, to check whether blocks and transactions arrive in the other node, but that means I'll have to update the functional_tests/README.md

Another option I see is turning the initialization of m_synchronized(offline) into m_synchronized(offline||regtest), which seems bad cause we might want to test for synchronization

moneromooo-monero commented:
You could also add a --fixed-difficulty N flag to enable the "return 1" (which would be "return N"), so you can also test mining.
About --offline, I think it's OK to either use --offline or --add-exclusive-peer, depending on what you intend to do.

stoffu commented:
@vicsn Ah, that idea didn't come to my mind. I've made another update to my branch adding the `--fixed-difficulty` parameter and doing mining within the `generateblocks` RPC and update to README.

https://github.com/stoffu/monero/commit/3251fd0c22e3f1e833918c6a3063694bb48225a1

vicsn commented:
Thanks!

I wanted to build in cryptonote_core.cpp the following to let regtest fallback on offline:
```
m_exclusive_nodes = command_line::get_arg(vm, nodetool::arg_p2p_add_exclusive_node).empty();
if(regtest && !exclusive_nodes) m_offline = true;

```
However, in order to access arg_p2p_add_exclusive_node, cryptonote_core.cpp would have to include net_node.h, whilst net_node.inl already includes cryptonote_core.h

If this cyclical dependency is difficult to solve, I'm happy leaving the fallback out, as the README is clear.

stoffu commented:
@vicsn 
I don’t think it’s necessary to include net_node.h; you can just declare arg_p2p_add_exclusive_node with `extern` and use it in crypronote_core.cpp.

Also, I intended to have the additional patch with --fixed-difficulty squashed into the main commit adding --regtest. I think it makes sense.

vicsn commented:
I squashed the commit :)

In the meantime I've been trying to make the ```command_line::arg_descriptor``` take both ```command_line::arg_descriptor<bool> arg_offline``` and ```const command_line::arg_descriptor<std::vector<std::string> > arg_p2p_add_exclusive_node``` as arguments. 

To achieve this, I would have to change the ```std::array``` in common/commandline.h (which can only take one type) into a ```std::tuple```. Do you think this change is 1) correct? 2) worth it?

moneromooo-monero commented:
tbh I think we should not make too many magic arguments. It can get confusing. If you want offline, use --offline, etc. Not a huge chore.

moneromooo-monero commented on src/cryptonote_core/blockchain.cpp:
```diff
@@ -351,6 +351,11 @@ bool Blockchain::init(BlockchainDB* db, const network_type nettype, bool offline
+  if (m_regtest)
+  {
+    m_fixed_difficulty = *regtest_diff;
+  }
```
That seems unfortunate. I often hack a "return constant" for difficulty on testnet. It's a useful thing to have, and this new change limits it to regtest mode.

stoffu replied:
OK, I made this patch to accommodate your request: https://github.com/stoffu/monero/commit/1f283ac93a12b8dc910cdc270c0ba15fe4471126

moneromooo-monero commented on src/cryptonote_core/cryptonote_core.cpp:
```diff
@@ -508,7 +531,16 @@ namespace cryptonote
         blocks_per_sync, sync_mode, fast_sync);
 
-    r = m_blockchain_storage.init(db.release(), m_nettype, m_offline, test_options);
+    const std::pair<uint8_t, uint64_t> regtest_hard_forks[3] = {std::make_pair(1, 0), std::make_pair(7, 1), std::make_pair(0, 0)};
```
That's going to be forgotten when we switch to v8, etc. If you want to set it to the last fork, request it from the hard fork object (get_ideal_version).

vicsn replied:
Do you mean move the logic to ```cryptonote_core.cpp``` to ensure it's close to the other hard_fork structs and it won't be forgotten? I don't see how ```get_ideal_version``` comes into play here, as for the regtest/FAKECHAIN nettype hardforks still need to be set for a low height...

stoffu replied:
@moneromooo-monero 
The function `get_ideal_version` assumes that the internal hard fork heights info is already initialized, which isn't the case here. Also, in this case we want to get the hard fork info heights info of MAINNET while the network type is set to FAKECHAIN. I've made a patch https://github.com/stoffu/monero/commit/a758f1dc7a878f086e36b5101fe0987e548a9213 that makes `HardFork:: Params` a public class and adds a static function `Blockchain::get_hard_fork_heights(network_type nettype)`.

moneromooo-monero replied:
Ah, yes, chicken and egg. OK. Somehow it seems like we're going about this the wrong way, but I can't quite put my finger on it.

vicsn replied:
Updated with stoffu's adjustments. I've tried thinking about what would be wrong about the initialization of regtest_hard_forks, but can't seem to identify the chicken and egg situation here...

stoffu commented on tests/functional_tests/speed.py:
```diff
@@ -0,0 +1,87 @@
+            while wallet.get_balance()['unlocked_balance'] == 0:
+                print('Waiting for wallet to refresh...')
+                sleep(1)
+        self._test_speed_transfer_split(wallet=wallet)
```
This line needs to be indented one level further; otherwise it gets executed only once. See my patch: https://github.com/stoffu/monero/commit/3649e7e12ba3a68e5b94efe1216c97a7fbcfc91d

stoffu commented:
The patch "update get_info RPC and bump RPC version" 0579b60 seems rather irrelevant for the regtest itself at this point, although I'm fine with adding the string field for its own sake. However, the patch seems incomplete because the corresponding change in core_rpc_server.cpp is not made.

vicsn commented:
I added nettype to get_info RPC call.
I also added more functional test cases in ```blockchain.py```, not for the sake of being complete, but to ensure I didn't overlook more in the current PR. As a result, I noticed RPC server still thought it was on mainnet. I therefore added a check for regtest in ```daemon.cpp```

moneromooo-monero commented:
Come to think of it, does that wallet manage to talk to that daemon without a --regtest of its own ? Or is the wallet deemed to be out of scope here?

vicsn commented:
In my original PR I added --regtest to the wallet as well because of the different config parameters. But in the current state it works fine as a mainnet wallet, I don't see the need to add --regtest anymore

moneromooo-monero commented:
OK, thank for confirming.
